### ISSUE TITLE
Segment boundaries are systematically late at speech_pad_ms=0 (onset P50 ≈ +40 ms with a heavy tail to +1.4 s, offset P50 ≈ +50 ms); the two ends need corrections in opposite directions, so speech_pad_ms cannot compensate

### ISSUE THREAD

**Issue by LXP-Never** [snakers4/silero-vad#798]
# Summary

When tagging keyword-spotting training data (multilingual headset recordings, 16 kHz), we measured `get_speech_timestamps` (non-streaming, ONNX) boundaries against an objective, annotation-free acoustic reference on **8,966 files**. At the library-default-style setting `speech_pad_ms=0`:

| boundary | lag vs. objective acoustic boundary (P50) | P90 | max |
|---|---|---|---|
| onset  | **+39 to +43 ms** (98-99% of files positive) | +130 to +140 ms | **+1036 to +1415 ms** |
| offset | **+47 to +58 ms** | +79 to +94 ms | — |

Both boundaries are late, but they need **opposite** corrections: recovering the onset requires extending the segment *earlier* (positive pad), while recovering the offset requires trimming the segment *earlier* (negative pad). `speech_pad_ms` moves both ends outwards symmetrically, so it can fix the onset only by making the offset error grow 1:1 (verified below: pad=50 ms reproduces a manual onset annotation within 4 samples while pushing the offset +49 ms further out).

*Edit note (2026-09-18): the first version of this report defined the acoustic reference as the first/last envelope frame above floor+6 dB, which single-frame clicks contaminated (inflated onset P50 to ~+79 ms and produced a spurious heavy tail). All numbers above use the click-robust definition (first/last frame of a run of ≥3 consecutive frames, i.e. 15 ms, above floor+6 dB). Figures and the example audio are attached in a comment below.*

# Environment

- silero-vad: `torch.hub.load('snakers4/silero-vad', model='silero_vad', onnx=True)`, master branch cache (model file `silero_vad.onnx`)
- Python 3.11, PyTorch 2.1.1+cu121 (CPU inference)
- Params: `threshold=0.5, window_size_samples=512, speech_pad_ms=0`, other args at library defaults
- Audio: 16 kHz mono (one channel extracted from multichannel headset recordings), preprocessed with a 100 Hz high-pass and peak normalization (standard VAD preprocessing; the bias persists without it)

# Single-file evidence (manual ground truth available)

One zh command utterance ("上一曲", 1.85 s; mono extract attached in the comment below). Manual annotated onset = sample 7908; objective energy onset (click-robust) = 8160; energy offset = 21520.

| threshold | speech_pad_ms | window | detected onset | onset vs. manual GT | detected offset | offset vs. energy offset |
|---|---|---|---|---|---|---|
| 0.5 | 0   | 512  | 8704 | **+796 (+50 ms)** | 21504 | −1 ms |
| 0.5 | 30  | 512  | 8224 | +316 (+20 ms) | 21984 | +29 ms |
| 0.5 | 50  | 512  | 7904 | **−4 (≈ GT)** | 22304 | **+49 ms** |
| 0.3 | 0   | 512  | 8704 | +796 (threshold does not move onset) | 22016 | +31 ms |
| 0.2 | 0   | 512  | 8192 | +284 | 22528 | +63 ms |
| 0.5 | 0   | 1024 | 8704 | +796 (window size does not move onset) | 21504 | −1 ms |
| 0.5 | 0   | 1536 | 8704 | +796 | 21504 | −1 ms |

Observations:
1. Lowering `threshold` or changing `window_size_samples` does not reliably recover the onset; only `speech_pad_ms` shifts it, and it shifts the offset outwards by the same amount (offset error grows from −1 ms to +49 ms as pad goes 0→50 ms).
2. For reference, FireRedVAD (DFSMN, FireRedTeam/FireRedVAD) on the same file errs in the opposite direction at the onset (7040, −54 ms) while also over-extending the offset (24160, +164 ms), i.e. boundary placement is model-specific and worth improving on both sides.

# Large-scale evidence (n = 8,966, annotation-free)

Methodology: per file we compute a 5 ms RMS envelope of the preprocessed signal and estimate the noise floor as the 30th percentile of frame dB levels. The acoustic onset/offset are the first/last frame of a run of **≥3 consecutive frames (15 ms)** above floor+6 dB; the run-length requirement rejects single-frame clicks (without it, a −45 dB one-frame click at sample 3200 of the example file masquerades as the onset). This reference is conservative: on the manually annotated file above it yields 8160 vs. manual 7908, i.e. the proxy is itself ~16 ms later than a human annotator, so the reported onset lag is if anything underestimated.

Results (zh headset commands, two loudness classes):

| class | n | onset lag P50 | onset lag P90 | onset lag max | offset lag P50 | offset lag P90 |
|---|---|---|---|---|---|---|
| comfortable | 300 | +43 ms | +140 ms | +1036 ms | +47 ms | +79 ms |
| soft | 8,666 | +39 ms | +130 ms | +1415 ms | +58 ms | +94 ms |

Onset lag is positive in 98-99% of files. The onset heavy tail (up to +1.4 s) corresponds to files where a weak initial syllable never raises the frame probability above threshold; examples available on request. (Detection recall itself is fine: only 5/8,671 soft files yielded no segment, so this report is not about misses.)

# Minimal reproduction

```python
import torch, soundfile as sf
model, utils = torch.hub.load('snakers4/silero-vad', model='silero_vad', onnx=True)
(get_speech_timestamps, _, _, _, _) = utils
wav, sr = sf.read("silero_issue_example_mono16k.wav", dtype="float32")  # attached in comment

for pad in (0, 30, 50):
    ts = get_speech_timestamps(wav, model, sampling_rate=sr,
                               threshold=0.5, window_size_samples=512,
                               speech_pad_ms=pad)
    print(pad, ts[0]["start"], ts[-1]["end"])
# 0  -> 8704 21504   (onset +50 ms vs manual GT 7908)
# 30 -> 8224 21984
# 50 -> 7904 22304   (onset ≈ GT, offset +49 ms late)
```

# Questions for the maintainers

1. Is the systematic lateness of both raw boundaries (onset P50 ~+40 ms, offset P50 ~+50 ms at pad=0) known and intended? The onset in particular appears to be emitted where frame probability crosses the threshold on soft consonant onsets, without back-tracing to the acoustic onset.
2. Would you consider hysteresis-style back-tracing (extend segment start backwards while frame probability stays above a lower threshold, and trim the offset symmetrically), so that boundary placement no longer depends on `speech_pad_ms`, which can only move both ends outwards?
3. For boundary-accurate tagging use cases (KWS/ASR data cutting, subtitle alignment), is there a recommended parameter set other than symmetric `speech_pad_ms`?

# Why this matters

VAD timestamps are widely used to cut/label training data for KWS and ASR and for subtitle alignment. A systematic +40-50 ms shift trims initial consonants and appends decay tails to every cut segment, and because the two ends need opposite corrections, the documented compensation (`speech_pad_ms`) necessarily corrupts one end while fixing the other.

**Comment by LXP-Never** [comment]
## Visual evidence & repro audio

**Fig 1** — the single case from the table above: waveform (top) and 5 ms RMS envelope with noise floor (bottom); vertical lines = manual onset GT (7908, green) and click-robust energy onset/offset (8160/21520, gray); horizontal bars = detected segments of silero pad=0 ms (red), silero pad=50 ms (orange), FireRedVAD (blue). Note the pad=0 onset sitting ~50 ms after both the manual GT and the acoustic energy rise, while its offset is exact; pad=50 ms recovers the onset but over-extends the offset.

![fig1](https://raw.githubusercontent.com/LXP-Never/silero-vad-issue-assets/main/silero_issue_fig1_single_case.png)

**Fig 2** — boundary bias distribution over n=8,966 zh command files (click-robust reference): onset lag (left) and offset bias (right) at thr=0.5/pad=0/win=512, dashed lines = per-class P50. Both ends are late (onset P50 ≈ +40 ms with a tail to +1.4 s; offset P50 ≈ +50 ms), and the two ends need corrections in opposite directions.

![fig2](https://raw.githubusercontent.com/LXP-Never/silero-vad-issue-assets/main/silero_issue_fig2_stats.png)

**Example audio** — [silero_issue_example_mono16k.wav](https://raw.githubusercontent.com/LXP-Never/silero-vad-issue-assets/main/silero_issue_example_mono16k.wav): mono 16 kHz 16-bit PCM extract (headset channel LFF1 + 100 Hz high-pass + peak normalization) of the utterance used in the single-case table. Feeding it to `get_speech_timestamps(..., threshold=0.5, window_size_samples=512, speech_pad_ms=0)` reproduces onset=8704 / offset=21504 exactly, and the pad sweep in the reproduction snippet of the issue body.

**Comment by snakers4** [comment]
Hi,

Many thanks for your feedback.

> The onset heavy tail (up to +1.4 s) corresponds to files where a weak initial syllable never raises the frame probability above threshold; examples available on request.

Please send typical and the edge cases and / or your validation dataset in its entirety if possible for us to play with.

> non-streaming, ONNX
> model, utils = torch.hub.load('snakers4/silero-vad', model='silero_vad', onnx=True)

There just was a release of a truly "non-steaming" model, which is in fact a "bit-by-bit faithful" conversion of a basic streaming model (it's confusing, I know).
But I suppose you used the previous plain release (i.e. `v6.2.1`). 
**_In any case, the model itself IS STREAMING_**, we did not realease a truly non-streaming model yet for a number of reasons.
The truly non-streaming model can "look into the future" and has better metrics.

> window
> 1024 1536
> Lowering threshold or changing window_size_samples does not reliably recover the onset; only speech_pad_ms shifts it, and it shifts the offset outwards by the same amount (offset error grows from −1 ms to +49 ms as pad goes 0→50 ms).

Many many versions ago our VAD accepted variable window sizes.
But now this option is deprecated and the window is fixed.
This param should have no effect on the metrics, which is backed by this table in your post, which is correct.

> Lowering threshold or ...

Well, the thresholds ideally should be properly tuned on domain-per-domain basis.
The easiest way to do this is to invoke the VAD with the `visualize_probs` flag.
 
> <img width="847" height="148" alt="Image" src="https://github.com/user-attachments/assets/17cf752e-5281-46b0-bae8-c61373eecaa5" />

So, if I interpret your data correctly, on average (without padding) **the VAD needs at least 1-1.5 frames of actual speech to trigger**.
And also **at least 1.5-2 frames to detrigger.**
But it only makes sense, since the VAD is not majic, it cannot see the future, it processes the data frame by frame (~32 ms).
And it needs a bit more "time" to "think".

> Params: threshold=0.5, window_size_samples=512, speech_pad_ms=0, other args at library defaults
>     speech_pad_ms: int (default - 30 milliseconds) Final speech chunks are padded by speech_pad_ms each side
> only speech_pad_ms shifts it, and it shifts the offset outwards by the same amount (offset error grows from −1 ms to +49 ms as pad goes 0→50 ms).
> Both boundaries are late, but they need opposite corrections: recovering the onset requires extending the segment earlier (positive pad), while recovering the offset requires trimming the segment earlier (negative pad). speech_pad_ms moves both ends outwards symmetrically, so it can fix the onset only by making the offset error grow 1:1 (verified below: pad=50 ms reproduces a manual onset annotation within 4 samples while pushing the offset +49 ms further out).

Well, the default param and your table kind of answer the issue. Since the VAD statisticall (great analysis by the way!) error is about 1 - 1.5 frames on the left and about 1.5-2 frames on the right, it can be easily adjusted accordingly, which we did in the default param. But we did not really think twice about the direction of the padding. Initially, many years ago we just slapped the symmetrical padding and that was it.

We missed that it we probably should move the offset to the left, not to the right. But this should be a very easy fix!

> Methodology: per file we compute a 5 ms RMS envelope of the preprocessed signal and estimate the noise floor as the 30th percentile of frame dB levels. The acoustic onset/offset are the first/last frame of a run of ≥3 consecutive frames (15 ms) above floor+6 dB; the run-length requirement rejects single-frame clicks (without it, a −45 dB one-frame click at sample 3200 of the example file masquerades as the onset). This reference is conservative: on the manually annotated file above it yields 8160 vs. manual 7908, i.e. the proxy is itself ~16 ms later than a human annotator, so the reported onset lag is if anything underestimated.

This is a nice baseline, but I guess it would fail on more noisy data and work similar to WebRTC VAD and would not be robust to loud non-speech sounds?

> i.e. the proxy is itself ~16 ms later than a human annotator, so the reported onset lag is if anything underestimated.

If you have human labels for some validation data, can you please share it privately, we will be happy to use it for next releases as +1 benchmark.

> Is the systematic lateness of both raw boundaries (onset P50 ~+40 ms, offset P50 ~+50 ms at pad=0) known and intended? The onset in particular appears to be emitted where frame probability crosses the threshold on soft consonant onsets, without back-tracing to the acoustic onset.

This is by design, because our VAD is a streaming one. It needs at least one frame to react.
Hence the default `speech_pad_ms: int (default - 30 milliseconds)` is set to about 30ms. Probably we should increase it to 40-50ms. 
We have decided not to publish our non-streaming one yet.
When [comparing ](https://github.com/snakers4/silero-vad/wiki/Quality-Metrics) different VADs, non-steaming VADs get an advantage, because they can "see the future". But even our **steaming** VAD has the best overall metics.
So we left it as it is now.

> Would you consider hysteresis-style back-tracing (extend segment start backwards while frame probability stays above a lower threshold, and trim the offset symmetrically), so that boundary placement no longer depends on speech_pad_ms, which can only move both ends outwards?

Not sure that I understand the exact algorithm proposition here, but we have something similar already implemented. 
The VAD session is "sticky", e.g. [here](https://github.com/snakers4/silero-vad/blob/60b7ffa243625ebdc1070275a29f18c87843786a/src/silero_vad/utils_vad.py#L406-L407) and also some small silence is tolerated to suppress some sneaky short silences.

> For boundary-accurate tagging use cases (KWS/ASR data cutting, subtitle alignment), is there a recommended parameter set other than symmetric speech_pad_ms?

I see a much easier staright-forward solution.
We should pad the onset of the speech chunks to "the left" (as it works now) and the offset to "the left" as well (now it is just padded "to the right", i.e. symmetrically).

> Visual evidence & repro audio

We would be happy if you provided a notebook for your visualizations / tests, so the we could test  your method on other / noisier data and compare it with your manually annotated data on our internal datesets.

Many thanks!
